app/keepers/keepers.go

```diff
@@ -226,7 +247,7 @@ func (appKeepers *AppKeepers) InitNormalKeepers(
 	// Create static IBC router, add transfer route, then set and seal it
 	ibcRouter := porttypes.NewRouter()
 	ibcRouter.AddRoute(icahosttypes.SubModuleName, icaHostIBCModule).
-		AddRoute(ibctransfertypes.ModuleName, transferIBCModule)
+		AddRoute(ibctransfertypes.ModuleName, &rateLimitingTransferModule)
```

Comment:
I'm confused about this wiring. We make our middleware the transfer module, but we pass the middleware into the transfer keeper as well?

Reply:
yeah, this is confusing, but that's how you wire up the middlewares. What we pass into the keeper is not the middleware, but the ICS4Wrapper which would be the middleware's keeper if we had one (Notice that the transfer keeper has two params that originally pointed to the channel: ics4Wrapper and channelKeeper; we override the first one)

So the ibc module's methods (the OnSomething methods) all go to the part implementing the IBCModule interface, but the ICS4 methods (SendPacket and WriteAck) go to the ICS4Wrapper (which ends up in the channel). This is because things go in one direction on send and the other on receive. So we get:

SendPacket -> Transfer -> RateLimit -> channel

OnRecv -> RateLimit -> transfer -> channel


But I'll double check this to make sure I'm wrapping it correctly

Reply:
hrmm, I somewhat think I get it? So middleware from IBC is currently insufficient to do everything we need, since we want rate limit code before other code in the receive code path?

(Not sure why it couldn't be in transfer first though, since as long as we error anywhere, it should get atomically reverted)

Reply:
No, this is not a workaround to get our use case to work; this is unfortunately the way the middleware wiring us designed to be used. The transfer IBCModule doesn't implement SendPacket or WriteAcknowledgement, as it only [implements](https://github.com/cosmos/ibc-go/blob/main/modules/apps/transfer/module.go#L29) porttypes.IBCModule and not porttypes.ICS4Wrapper ([definitions](https://github.com/cosmos/ibc-go/blob/main/modules/core/05-port/types/module.go)). It instead relies on the transfer module's keeper to have an [ICS4Wrapper](https://github.com/cosmos/ibc-go/blob/main/modules/apps/transfer/keeper/keeper.go#L23-L24) that is [initialized](https://github.com/cosmos/ibc-go/blob/main/testing/simapp/app.go#L423) to either the channelKeeper (which implements ICS4Wrapper) or the ICS4Wrapper implementation of the middleware.

The middleware [interface](https://github.com/cosmos/ibc-go/blob/main/modules/core/05-port/types/module.go#L132) from IBC requires users to implement both porttypes.IBCModule and porttypes.ICS4Wrapper, but the later can't really wrap SendPacket or WriteAcknowledgement because they don't exist in the transfer ibc module. 

I think the reason for designing it this way is so that a middleware would behave the same way as if it was a function wrapping another: what originates from the outside (other chain) must flow inwards and what originates in the last function called (our chain) must flow outwards. 

It's probably easier if I go over the code in a call though, so let me know so we can screenshare